cpg_infra/abstraction/hailbatch.py

```diff
@@ -0,0 +1,212 @@
+                f'Could not close billing project {props["name"]}: {resp.text}'
+            )
+
+    def diff(self, _id, old_inputs, new_inputs):
+        replaces = []
+        if old_inputs['name'] != new_inputs['name']:
+            replaces.append('name')
+
+        if old_inputs['batch_uri'] != new_inputs['batch_uri']:
+            replaces.append('batch_uri')
+
+        return pulumi.dynamic.DiffResult(
+            len(replaces) > 0, replaces, stables=[], delete_before_replace=False
+        )
+
+
+class HailBatchBillingProjectMembershipProvider(pulumi.dynamic.ResourceProvider):
+    """Pulumi provider for membership to a Hail Batch Billing Project"""
+
+    def create(self, props) -> pulumi.dynamic.CreateResult:
+        billing_project = props['billing_project']
+
+        if isinstance(billing_project, HailBatchBillingProject):
+            billing_project_id = billing_project.id
+        else:
+            billing_project_id = billing_project
+
+        user = props['user']
+
+        token_category, batch_uri, billing_project_name = billing_project_id.split('::')
+        url = HAIL_ADD_USER_TO_BILLING_PROJECT_PATH.format(
+            hail_batch_url=batch_uri,
+            billing_project=billing_project_name,
+            user=user,
+        )
+        hail_auth_token = get_hail_batch_token(token_category)
+        resp = requests.post(
+            url, headers={'Authorization': f'Bearer {hail_auth_token}'}, timeout=60
+        )
+        resp.raise_for_status()
+
+        return pulumi.dynamic.CreateResult(
+            id_=f'{token_category}::{batch_uri}::{billing_project_name}::{user}',
+            outs=props,
+        )
+
+    def diff(self, _id: str, _olds, _news) -> pulumi.dynamic.DiffResult:
+        replaces = []
+
+        if _olds['billing_project'] != _news['billing_project']:
+            replaces.append('billing_project')
+        if _olds['user'] != _news['user']:
+            replaces.append('user')
+
+        return pulumi.dynamic.DiffResult(
+            changes=len(replaces) > 0,
+            replaces=replaces,
+            delete_before_replace=len(replaces) > 0,
+        )
+
+    def delete(self, _id: str, _props) -> None:
+        bp_components = _props['billing_project'].split('::')
+        token_category, batch_uri, billing_project_name = bp_components
+        url = HAIL_DELETE_USER_FROM_BILLING_PROJECT_PATH.format(
+            hail_batch_url=batch_uri,
+            billing_project=billing_project_name,
+            user=_props['user'],
+        )
+
+        hail_auth_token = get_hail_batch_token(token_category)
+        resp = requests.post(
+            url, headers={'Authorization': f'Bearer {hail_auth_token}'}, timeout=60
+        )
+
+        if not resp.ok:
+            raise ValueError(f'Could not delete user from billing project: {resp.text}')
+
+    def read(self, id_: str, props) -> pulumi.dynamic.ReadResult:
+        # TODO: implement read from hail batch to determine if still present
```

Comment:
Didn't you mention that you implemented this? Did you already commit that?

Reply:
Implemented now